### PR TITLE
Refactor pseudo code part 2

### PR DESCRIPTION
**Pull request by ludugz**
### Description
- Update branch `refactor-psuedo-code` to run on `banner` & `item` cases
- For other cases, will be handled in the next PR.

### Demo
(Notice that the banners are displayed properly with different types of URL)
Mar-09-2022 22-37-09

### CODE REVIEW

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt [hunk 1]
```diff
@@ -1,31 +1,60 @@
...
 
-    fun createViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
+    fun createViewHolder(viewType: Int, parent: ViewGroup): BaseViewHolder {
+        if (viewType == INVALID) {
+            // TODO: INVALID case should be refactored later
+            val view = LayoutInflater.from(parent.context)
+                .inflate(R.layout.empty_view_row_item, parent, false)
+            return EmptyViewHolder(view)
+        }
         return factories[viewType].createViewHolder(parent)
```
Comment (ludugz) [inline weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt]: I encountered an index exception where the list index is passed as `-1` (because the `INVALID` value is `-1`). Made a quick fix for this here.

**Review by ludugz — COMMENTED** [review]
Left 1 comment for explanation.

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt (hunk 1 above)
Comment (ddikman) [reply weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt]: Good call 👍

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt [hunk 2]
```diff
@@ -1,31 +1,60 @@
...
     }
 
     companion object {
-        const fallback: int = -1;
+        const val INVALID = -1
     }
 }
 
-class FactoryMethod(type: String, factory: (parent: ViewGroup) -> BaseViewHolder) {
+class FactoryMethod(val type: String) {
```
Comment (ddikman) [inline weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt]: nope :) this is what we want to avoid. use inheritance or lambdas to register these. avoid switch cases

https://blog.cleancoder.com/uncle-bob/2021/03/06/ifElseSwitch.html

https://www.linkedin.com/pulse/why-switch-statements-bad-th%C3%A9o-farnole-/?trk=public_profile_article_view

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/ShopListViewHolder.kt [hunk 3]
```diff
@@ -14,12 +14,12 @@ class ShopListViewHolder(
 ) : BaseViewHolder(itemView) {
     private val title: TextView = itemView.findViewById(R.id.text_view_shop_name)
     private val recyclerView: RecyclerView = itemView.findViewById(R.id.recycler_view_shop)
-    override fun bindData() {
+    override fun bindData(data: WidgetData) {
         widgetDataList?.forEach { widgetData ->
             if (widgetData.type == "shop_list") {
                 val data = widgetData.data
```
Comment (ddikman) [inline weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/ShopListViewHolder.kt]: you already have the data in the parameter so this all needs to go

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/ItemListViewHolder.kt [hunk 4]
```diff
@@ -6,25 +6,41 @@ import androidx.recyclerview.widget.LinearLayoutManager
...
+    override fun bindData(data: WidgetData) {
+        recyclerView.apply {
+            // TODO: Left for future reference, but can probably be deleted
+//            val itemList = data.data["items"] as List<BaseItem>
+//            val itemList = (data.data["items"] as List<*>).filterIsInstance<ItemListItem>().takeIf {
+//                it.size == (data.data["items"] as List<*>).size
+//            }
+            
+            // TODO: Might need to refactor this part
+            val itemList = (data.data["items"] as List<*>?)?.map {
```
Comment (ddikman) [inline weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/ItemListViewHolder.kt]: 👍 yep, this would be a good way to parse the data inside the viewholder itself

break it out to a method to make it even easier to read what is parsing logic and what is rendering logic

**Review by ddikman — COMMENTED** [review]
Added some comments

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/ItemListViewHolder.kt (hunk 4 above)
Comment (ludugz) [reply weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/ItemListViewHolder.kt]: Fixed in https://github.com/ludugz/weld-rendering/pull/2/commits/d08c902b0f2959b2eb3e0fa9cb437ca0c49910c6

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt (hunk 2 above)
Comment (ludugz) [reply weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt]: Thanks for the article, changed to use lambda to register in `MainActivity`

File: weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt (hunk 1 above)
Comment (ludugz) [reply weld-rendering/app/src/main/java/styler/weld/rendering/viewholder/WidgetFactory.kt]: Refactored invalid type in here https://github.com/ludugz/weld-rendering/pull/2/commits/e3f7636caa3fd24e4ea7c60b5149e57ff0e3082e

File: weld-rendering/app/src/main/java/styler/weld/rendering/adapter/HorizontalAdapter.kt [hunk 5]
```diff
@@ -13,7 +13,7 @@ import styler.weld.rendering.viewholder.HorizontalShopViewHolder
 import styler.weld.rendering.viewholder.HorizontalViewHolder
 
 class HorizontalAdapter(
-    private val baseList: List<BaseItem>?,
+    private val baseList: List<HorizontalListItem>?,
     private val type: String
 ) :
```
Comment (ludugz) [inline weld-rendering/app/src/main/java/styler/weld/rendering/adapter/HorizontalAdapter.kt:18]: Similar to `WidgetAdapter`, we should need a Factory class here (Will do it in another PR)

File: weld-rendering/app/src/main/java/styler/weld/rendering/factory/FactoryMethod.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package styler.weld.rendering.factory
+
+import android.view.ViewGroup
+import styler.weld.rendering.viewholder.BaseViewHolder
+
+class FactoryMethod(val type: String, val factory: (parent: ViewGroup) -> BaseViewHolder) {
+    fun createViewHolder(parent: ViewGroup): BaseViewHolder = factory(parent)
+}
```
Comment (ludugz) [inline weld-rendering/app/src/main/java/styler/weld/rendering/factory/FactoryMethod.kt:8]: I extracted this `FactoryMethod` in case there is another Factory that might use this class.

**Review by ludugz — COMMENTED** [review]
Already addressed all your comments, please let me know if there is something incorrect with what you meant.
Besides, left some comments for explanation.

File: weld-rendering/app/src/main/java/styler/weld/rendering/MainActivity.kt [hunk 7]
```diff
@@ -1,56 +1,80 @@
...
+        }
 
-        // examples in case you would handle area and feed in the same
-//        widgetFactory.register("area_selection", ....factory)
-//        widgetFactory.register("feed", ....feedfactory)
-//        widgetData.add(0, new WidgetData(type: "area_selection"))
-//        widgetData.add(widgetData.length - 1, new WidgetData(type: "feed"))
+        widgetFactory.register("shop_list") { parent ->
+            val view = LayoutInflater.from(parent.context)
+                .inflate(R.layout.shop_list_row_item, parent, false)
```
Comment (ddikman) [inline weld-rendering/app/src/main/java/styler/weld/rendering/MainActivity.kt]: this part here `R.layout.shop_list_row_item` should be owned by the ShopListViewHolder. this method here, that you do as a lamba, can you add it in or near the article list view holder?

I'm thinking I would expect something like:

```
widgetFactory.register("article_list") { parent -> ArticleListViewHolder.create(parent) }
```

Comment (ludugz) [reply weld-rendering/app/src/main/java/styler/weld/rendering/MainActivity.kt]: Yeah agree that should handle traits inside of each class, makes sense that we can move each `R.id` handling to view holder. 
Fixed in https://github.com/ludugz/weld-rendering/pull/2/commits/81001fa5ebaac0aa34abbc7f2ef58feac559b485
